syz-manager/repro.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"context"
+	"maps"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/syzkaller/pkg/instance"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/stats"
+	"github.com/google/syzkaller/vm"
+)
+
+type repoManagerView interface {
+	// TODO: consider moving the function to reproManager.
+	runRepro(crash *Crash) *ReproResult
+}
+
+type reproManager struct {
+	Pending            atomic.Int64
+	StatNumReproducing *stats.Val
+	Done               chan *ReproResult
+
+	pool *instance.Pool[*vm.Instance]
+
+	onlyOnce bool
+
+	mu          sync.Mutex
+	ctx         context.Context
+	mgr         repoManagerView
+	reproducing map[string]bool
+	attempted   map[string]bool
+	parallel    chan struct{}
+}
+
+func newReproManager(ctx context.Context, pool *instance.Pool[*vm.Instance],
+	onlyOnce bool, mgr repoManagerView) *reproManager {
+	ret := &reproManager{
+		StatNumReproducing: stats.Create("reproducing", "Number of crashes being reproduced",
+			stats.Console, stats.NoGraph),
+		Done: make(chan *ReproResult, 10),
+
+		ctx:         ctx,
+		pool:        pool,
+		onlyOnce:    onlyOnce,
+		mgr:         mgr,
+		parallel:    make(chan struct{}, pool.Total()),
+		reproducing: map[string]bool{},
+		attempted:   map[string]bool{},
+	}
+	go ret.adjustPoolSize()
+	return ret
+}
+
+// startReproduction() is assumed to be called only once.
+// The agument is the maximum number of VMs dedicated to the bug reproduction.
+func (m *reproManager) startReproduction(VMs int) {
+	log.Logf(1, "starting reproductions (%d VMs available)", VMs)
+	for count := 1; count <= VMs; count++ {
```

Comment:
This can be:
```
	for count := 1; m.needVMs(count) <= VMs; count++ {
		m.parallel <- struct{}{}
	}
```

But also if I have only 1 VM, or 2 VMs + 1 dedicated for fuzzing, we still need to run at least 1 repro. If I am reading the code correctly, it won't run any repros at all.

Reply:
Good point!